### DISCUSSION TITLE
Fullstack .NET - Testando a API - Create, Update e Delete - converter DateTime para DateOnly

### DISCUSSION THREAD

**Discussion by MelloRenato** [balta-io/forum#672] [Tira Dúvidas] [answered]
Prezados, 

Estou utilizando o MySql ao invés do MS-SQL.

Ao definir meu modelo, opter por utilizar o tipo DateOnly para a data.

Criei as tabelas via EF e executei o request create sem problemas.

Ao teste as requisições Delete e Update ocorreu um erro "InvalidCastException" (segue imagem abaixo).

Como resolvo esse problema? 

![ErroCast](https://github.com/user-attachments/assets/71f066e2-416f-4d46-909f-6a5b2c4b4ed1)

Grato,
Renato

**Comment by henriqueweege** [comment, marked as answer]
Oi, @MelloRenato! Tudo certo?
Então, para conseguir te ajudar precisaria de um pouco mais de contexto.
Existe algum tipo de conversão de tipos entre a tua request e o model?
Como estava tua request?
Ao criar o mapeamento dos campos no C# para realizar as migrations, você utilizou algum tipo específico para os campos de data?

Valeeeu!

**Reply by andrebaltieri** [reply]
Você é fera 🚀

**Comment by MelloRenato** [comment]
Olá Henrique,
Não existe nenhuma conversão.
O modelo está definido da seguinte forma
public DateOnly? DataNasc { get; set; } = null;public DateOnly DataCadastro { get; set; } = DateOnly.FromDateTime(DateTime.Now);
No request, está da mesma forma e o banco, MySQL, como Date.
No migration está desta forma:DataNasc = table.Column<DateOnly>(type: "Date", nullable: true),DataCadastro = table.Column<DateOnly>(type: "Date", nullable: false),  
Se você preferir posso subir tudo para o GitHub.
Abs.
Renato

    Em segunda-feira, 26 de agosto de 2024 às 08:29:49 BRT, Andre Baltieri ***@***.***> escreveu:  
 
 


Você é fera 🚀

**Reply by henriqueweege** [reply]
@MelloRenato, eu não consegui reproduzir o erro na minha máquina : / teria como subir seu código para eu dar uma olhada?
Valeeeu!

**Reply by MelloRenato** [reply]
Olá @henriqueweege.

Desculpe pela minha ausência.

Segue o link do repositósio: https://github.com/MelloRenato/Consultorio_repo.git

Obrigado.

**Reply by henriqueweege** [reply]
Oi, @MelloRenato ! Tudo joia?
Tranquilo!! Dei uma olhada aqui, e pesquisando descobri que o mapeamento do Entity não funciona legal para DateOnly mesmo, então é necessário adicionar um conversor para esse tipo. 

Basicamente, você vai precisar criar essa classe aqui:
```cs
public class DateOnlyConverter : ValueConverter<DateOnly, DateTime>
{
    public DateOnlyConverter() : base(
        d => d.ToDateTime(TimeOnly.MinValue),
        d => DateOnly.FromDateTime(d))
    { }
}
```

e adicionar esse código a tua classe AppDbContext:
```
    protected override void ConfigureConventions(ModelConfigurationBuilder builder)
    {
        builder.Properties<DateOnly>()
                .HaveConversion<DateOnlyConverter>()
                .HaveColumnType("date");
    }
```

![image](https://github.com/user-attachments/assets/571cdaa4-9543-4893-8088-f1c6d0feda00)




e ai deve funcionar como esperado:
![image](https://github.com/user-attachments/assets/c45710a3-e4b8-44b4-a83c-110bd5a4ecb4)


Se quiser, você pode ler mais sobre [nesse artigo](https://code-maze.com/csharp-map-dateonly-timeonly-types-to-sql/) e [nesse post do StackOverflow](https://stackoverflow.com/questions/69146423/date-only-cannot-be-mapped-sql-server-2019).

Valeeeeeu!

**Reply by MelloRenato** [reply]
Oi @henriqueweege,

Primeiramente obrigado pela ajuda.

Vou implementar essas classes para ver se resolve, mas após ler o artigo fiquei com um ponto de "?".

De acordo com o artigo, o problema ocorre na versão 7 ou anterior do EF e que a partir da versão 8 isso é resolvido, ou seja, o EF 8 é com patível com o DateOnly, certo?

Eu estou usando o EF 8.0.8 e o .Net 8.0.403.

Outro ponto que me chamou a atenção, é que o artigo se refere ao MS-SQL e eu estou utilizando o MySQL.

Será que o DateOnly, mesmo na versão 8 do EF não é suportado para o MySQL?

De qualquer forma, vou implementar as classes e retorno assim que possível.

Abs.

**Reply by MelloRenato** [reply]
@henriqueweege,

Show!! Implementei as alterações e funcionou.

Então o EF 8 não trabalha legal com o DateOnly e MySQL.

Obrigado.

Abraço.
